File: ios/HRPayrollApp/Models/HRModels.swift
//
//  HRModels.swift
//  HRPayrollApp
//

import Foundation

public enum Department: String, CaseIterable, Codable, Identifiable {
    case engineering = "Engineering"
    case product = "Product"
    case design = "Design"
    case hr = "Human Resources"
    case finance = "Finance"
    case marketing = "Marketing"
    case operations = "Operations"
    
    public var id: String { rawValue }
}

public enum EmploymentType: String, CaseIterable, Codable {
    case fullTime = "Full-time"
    case partTime = "Part-time"
    case contract = "Contract"
    case intern = "Intern"
}

public struct JobPosition: Identifiable, Codable {
    public var id: UUID = UUID()
    public var code: String
    public var title: String
    public var department: Department
    public var minSalary: Double
    public var maxSalary: Double
    public var description: String
    public var status: String = "active"
}

public struct TaxInfo: Codable {
    public var taxId: String
    public var filingStatus: String // "Single", "Married"
    public var stateWithholdingRate: Double // e.g. 5.0
    public var federalAllowances: Int
    public var preTax401kPercent: Double
    public var monthlyHealthInsurance: Double
}

public struct Employee: Identifiable, Codable {
    public var id: UUID = UUID()
    public var employeeCode: String
    public var firstName: String
    public var lastName: String
    public var email: String
    public var phone: String
    public var photoBase64: String? = nil
    public var positionTitle: String
    public var department: Department
    public var joiningDate: Date
    public var probationMonths: Int = 3
    public var baseSalary: Double
    public var hourlyRate: Double
    public var isHourly: Bool = false
    public var taxInfo: TaxInfo
    public var bankAccount: String
    
    public var fullName: String {
        "\(firstName) \(lastName)"
    }
    
    public var tenureFormatted: String {
        let diff = Calendar.current.dateComponents([.year, .month], from: joiningDate, to: Date())
        let years = diff.year ?? 0
        let months = diff.month ?? 0
        if years > 0 {
            return "\(years)y \(months)m"
        }
        return "\(months) months"
    }
}

public struct WorkHourLog: Identifiable, Codable {
    public var id: UUID = UUID()
    public var employeeId: UUID
    public var date: Date
    public var regularHours: Double
    public var overtimeHours: Double
    public var holidayHours: Double
    public var isApproved: Bool
}

public struct PaySlip: Identifiable, Codable {
    public var id: UUID = UUID()
    public var slipNumber: String
    public var employeeName: String
    public var employeeCode: String
    public var positionTitle: String
    public var department: String
    public var payDate: Date
    public var grossPay: Double
    public var regularPay: Double
    public var overtimePay: Double
    public var federalTax: Double
    public var stateTax: Double
    public var socialSecurity: Double
    public var medicare: Double
    public var netPay: Double
}

File: android/app/src/main/java/com/hrpayroll/app/model/HRModels.kt
package com.hrpayroll.app.model

import java.time.LocalDate
import java.time.Period
import java.time.temporal.ChronoUnit
import java.util.UUID

enum class Department(val displayName: String) {
    ENGINEERING("Engineering"),
    PRODUCT("Product"),
    DESIGN("Design"),
    HR("Human Resources"),
    FINANCE("Finance"),
    MARKETING("Marketing"),
    OPERATIONS("Operations")
}

data class JobPosition(
    val id: String = UUID.randomUUID().toString(),
    val code: String,
    val title: String,
    val department: Department,
    val level: String = "Senior",
    val employmentType: String = "Full-time",
    val minSalary: Double,
    val maxSalary: Double,
    val description: String,
    val status: String = "active"
)

data class TaxInfo(
    val taxId: String,
    val filingStatus: String, // "Single", "Married"
    val stateWithholdingRate: Double, // percentage, e.g. 5.0
    val federalAllowances: Int = 1,
    val preTax401kPercent: Double = 4.0,
    val monthlyHealthInsurance: Double = 150.0
)

data class Employee(
    val id: String = UUID.randomUUID().toString(),
    val employeeCode: String,
    val firstName: String,
    val lastName: String,
    val email: String,
    val phone: String,
    val photoUrl: String? = null, // Photo avatar or image URI
    val positionTitle: String,
    val department: Department,
    val joiningDate: String, // YYYY-MM-DD
    val probationMonths: Int = 3,
    val baseSalary: Double,
    val hourlyRate: Double = 0.0,
    val isHourly: Boolean = false,
    val taxInfo: TaxInfo,
    val bankAccount: String
) {
    val fullName: String get() = "$firstName $lastName"

    // Calculates tenure from joining date
    fun getTenureDisplay(): String {
        return try {
            val join = LocalDate.parse(joiningDate)
            val now = LocalDate.now()
            val period = Period.between(join, now)
            when {
                period.years > 0 -> "${period.years}y ${period.months}m"
                period.months > 0 -> "${period.months}m ${period.days}d"
                else -> "${period.days} days"
            }
        } catch (e: Exception) {
            "Active"
        }
    }

    // Calculates probation status
    fun getProbationStatus(): String {
        return try {
            val join = LocalDate.parse(joiningDate)
            val probationEnd = join.plusMonths(probationMonths.toLong())
            val now = LocalDate.now()
            if (now.isBefore(probationEnd)) {
                val daysLeft = ChronoUnit.DAYS.between(now, probationEnd)
                "Probation ($daysLeft d left)"
            } else {
                "Confirmed Staff"
            }
        } catch (e: Exception) {
            "Confirmed"
        }
    }
}

data class WorkHourLog(
    val id: String = UUID.randomUUID().toString(),
    val employeeId: String,
    val employeeName: String,
    val date: String,
    val regularHours: Double,
    val overtimeHours: Double = 0.0,
    val holidayHours: Double = 0.0,
    val isApproved: Boolean = true,
    val notes: String = ""
)

data class PaySlip(
    val id: String = UUID.randomUUID().toString(),
    val slipNumber: String,
    val employeeId: String,
    val employeeName: String,
    val employeeCode: String,
    val positionTitle: String,
    val department: String,
    val payDate: String,
    val basePay: Double,
    val regularHoursWorked: Double,
    val regularPay: Double,
    val overtimeHoursWorked: Double,
    val overtimePay: Double,
    val holidayHoursWorked: Double,
    val holidayPay: Double,
    val allowances: Double,
    val grossPay: Double,
    val federalTax: Double,
    val stateTax: Double,
    val socialSecurity: Double,
    val medicare: Double,
    val preTax401k: Double,
    val healthInsurance: Double,
    val totalDeductions: Double,
    val netPay: Double,
    val bankSummary: String
)
